File: backend/models/gold_category.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// GoldCategory represents gold quality classification with pricing
type GoldCategory struct {
	ID          uint           `gorm:"primarykey" json:"id"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
	Code        string         `gorm:"not null;size:20" json:"code"` // e.g., "420", "750", "916", "999" - unique index created manually in migration
	Name        string         `gorm:"not null;size:50" json:"name"`             // e.g., "9K", "18K", "22K", "24K"
	Purity      *float64       `gorm:"default:null" json:"purity"`               // e.g., 0.375, 0.750, 0.916, 0.999 - nullable
	BuyPrice    float64        `gorm:"not null;default:0" json:"buy_price"`      // Harga beli per gram
	SellPrice   float64        `gorm:"not null;default:0" json:"sell_price"`     // Harga jual per gram
	Description string         `gorm:"size:255" json:"description"`
	IsActive    bool           `gorm:"default:true" json:"is_active"`
	Products    []Product      `gorm:"foreignKey:GoldCategoryID" json:"products,omitempty"`
}

File: mobile/lib/models/gold_category.dart
import 'package:equatable/equatable.dart';

class GoldCategory extends Equatable {
  final int id;
  final String code;
  final String name;
  final double? purity;
  final double buyPrice;
  final double sellPrice;
  final String description;
  final bool isActive;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  const GoldCategory({
    required this.id,
    required this.code,
    required this.name,
    this.purity,
    this.buyPrice = 0,
    this.sellPrice = 0,
    this.description = '',
    this.isActive = true,
    this.createdAt,
    this.updatedAt,
  });

  factory GoldCategory.fromJson(Map<String, dynamic> json) {
    return GoldCategory(
      id: json['id'] ?? 0,
      code: json['code'] ?? '',
      name: json['name'] ?? '',
      purity: json['purity']?.toDouble(),
      buyPrice: (json['buy_price'] ?? 0).toDouble(),
      sellPrice: (json['sell_price'] ?? 0).toDouble(),
      description: json['description'] ?? '',
      isActive: json['is_active'] ?? true,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      updatedAt: json['updated_at'] != null
          ? DateTime.parse(json['updated_at'])
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'name': name,
      'purity': purity,
      'buy_price': buyPrice,
      'sell_price': sellPrice,
      'description': description,
      'is_active': isActive,
    };
  }

  GoldCategory copyWith({
    int? id,
    String? code,
    String? name,
    double? purity,
    double? buyPrice,
    double? sellPrice,
    String? description,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return GoldCategory(
      id: id ?? this.id,
      code: code ?? this.code,
      name: name ?? this.name,
      purity: purity ?? this.purity,
      buyPrice: buyPrice ?? this.buyPrice,
      sellPrice: sellPrice ?? this.sellPrice,
      description: description ?? this.description,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  List<Object?> get props => [id, code, name, purity, buyPrice, sellPrice, isActive];
}
